lib/filterx/filterx-ref.c

```diff
@@ -55,7 +73,9 @@ _filterx_ref_setattr(FilterXObject *s, FilterXObject *attr, FilterXObject **new_
 
   _filterx_ref_cow(self);
 
-  return filterx_object_setattr(self->value, attr, new_value);
+  gboolean result = filterx_object_setattr(self->value, attr, new_value);
+  filterx_ref_propagate_root(*new_value, self);
```

Comment:
This implies that dirty tracking will not work purely on objects anymore, which is fine, but it made me think: object, ref, and dict got so coupled that certain functionality won't work properly without using the combinations of their API. `FilterXObject` should help with a facade in the future.

Reply:
True, dirty tracking can only work for objects wrapped in FilterXRefs. For now OTel classes are mutable but don't use the new API, but we discussed we would rework those and eliminate getattr/setattr for OTel objects.